src/components/EditableHeading/EditableHeading.tsx

```diff
@@ -0,0 +1,168 @@
+  weight?: HeadingWeight;
+  editable?: boolean;
+  onChange?: (value: string) => void;
+  focusOnMount?: boolean;
+  selectOnMount?: boolean;
+  ariaLabel?: string;
+}
+
+const EditableHeading: React.FC<EditableHeadingProps> & {
+  types?: typeof Heading.types;
+  weights?: typeof Heading.weights;
+} = forwardRef(
+  (
+    {
+      id,
+      className,
+      "data-testid": dataTestId,
+      value,
+      type = Heading.types.H1,
+      weight = HeadingWeight.NORMAL,
+      editable = true,
+      onChange = () => {},
+      focusOnMount = true,
+      selectOnMount = false,
+      ariaLabel = ""
+    },
+    ref
+  ) => {
+    const componentRef = useRef(null);
+    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });
+
+    const [isEditing, setIsEditing] = useState(false);
+    const [inputValue, setInputValue] = useState(value);
+    const [inputWidth, setInputWidth] = useState(0);
+
+    const inputRef = useRef(null);
+
+    function toggleEditMode() {
+      if (!editable || isEditing) {
+        return;
+      }
+
+      setIsEditing(true);
+    }
+
+    function handleInputValueChange() {
+      setIsEditing(false);
+      if (!inputValue || value === inputValue) {
+        setInputValue(value);
+        return;
+      }
+      onChange(inputValue);
+    }
+
+    function handleBlur() {
+      handleInputValueChange();
+    }
+
+    function handleKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
+      if (event.key === keyCodes.ENTER) {
+        handleInputValueChange();
+      }
+      if (event.key === keyCodes.ESCAPE) {
+        setIsEditing(false);
+        setInputValue(value);
+      }
+    }
+
+    function handleChange(event: React.ChangeEvent<HTMLInputElement>) {
+      setInputValue(event.target.value);
+    }
+
+    const clickableProps = useClickableProps(
+      {
+        onClick: toggleEditMode,
+        id,
+        role: "button"
+      },
+      mergedRef
+    );
+
+    function focus() {
+      if (inputRef.current) {
+        requestAnimationFrame(() => {
+          inputRef.current?.focus();
+        });
+      }
+    }
+
+    function select() {
+      if (inputRef.current) {
+        inputRef.current.select();
+      }
+    }
+
+    useLayoutEffect(() => {
+      if (focusOnMount) {
+        focus();
+      }
+      if (selectOnMount) {
+        select();
+      }
+    }, [focusOnMount, isEditing, selectOnMount]);
+
+    const inputClassNames = useMemo(() => {
+      return cx(styles.input, getStyle(styles, camelCase(type + "-" + weight)));
+    }, [type, weight]);
+
+    return (
+      <div
+        ref={mergedRef}
+        id={id}
+        aria-label={ariaLabel}
+        data-testid={dataTestId || getTestId(ComponentDefaultTestId.EDITABLE_HEADING, id)}
+        className={cx(styles.editableHeading, className)}
+      >
+        {isEditing && editable ? (
+          <>
+            <HiddenInputPlaceholder className={inputClassNames} value={inputValue} onChange={setInputWidth} />
```

Comment:
The main difference between this and the old one is that the input is now an inline element and not a black element, so since html inputs cannot have a dynamic width that fits the text we use this hack of a hidden placeholder that defines the width for the input. This is a very common hack and it's being used in libraries such as `react-input-autosize`.